p32/src/shapes.rs

```diff
@@ -0,0 +1,329 @@
+use std::fmt::Debug;
+
+#[derive(Eq, PartialEq, Debug)]
+pub enum Choice<T, U> {
+    First(T),
+    Second(U),
+}
+
+pub trait Shape {
+    const NAME: &'static str;
+
+    fn perimeter(&self) -> f32;
+    fn area(&self) -> f32;
+    fn scale(&mut self, factor: f32);
+    fn area_to_perimeter(&self) -> f32 {
+        match self.perimeter() {
+            0.0 => 0.0,
+            _ => self.area() / self.perimeter(),
+        }
+    }
+    fn biggest_shape<'a, 'b, T: Shape>(&'a self, other: &'b T) -> Choice<&'a Self, &'b T> {
+        if self.area() > other.area() {
+            Choice::First(self)
+        } else {
+            Choice::Second(other)
+        }
+    }
+    fn print_properties(&self) {
+        println!("Name: {}", Self::NAME);
+        println!("Area: {}", self.area());
+        println!("Perimeter: {}", self.perimeter());
+    }
+}
+
+#[derive(Clone, Debug, Copy, Default, PartialEq)]
+pub struct Point {
+    pub x: f32,
+    pub y: f32,
+}
+
+#[derive(Clone, Debug, Copy, Default, PartialEq)]
+pub struct Circle {
+    pub center: Point,
+    pub radius: f32,
+}
+
+#[derive(Clone, Debug, Copy, Default, PartialEq)]
+pub struct Triangle {
+    pub a: Point,
+    pub b: Point,
+    pub c: Point,
+}
+
+#[derive(Clone, Debug, Copy, Default, PartialEq)]
+pub struct Rectangle {
+    pub top_left: Point,
+    pub bottom_right: Point,
+}
+
+#[derive(Clone, Debug, Copy, PartialEq)]
+pub enum DynamicShape {
+    Point(Point),
+    Circle(Circle),
+    Triangle(Triangle),
+    Rectangle(Rectangle),
+}
+
+impl Shape for Point {
```

Comment:
You need to overwrite `area_to_perimeter` here. It should return 0 for `Point`, not INF.